sysapi/sysapi/GetCommandCode.c

```diff
@@ -41,7 +41,8 @@ TSS2_RC Tss2_Sys_GetCommandCode(
     if (ctx->previousStage == CMD_STAGE_INITIALIZE)
         return TSS2_SYS_RC_BAD_SEQUENCE;
 
-    *(TPM2_CC *)commandCode = HOST_TO_BE_32(ctx->commandCode);
+    TPM2_CC tmp = HOST_TO_BE_32(ctx->commandCode);
+    memcpy(commandCode, (void *)&tmp, sizeof(tmp));
```

Comment:
No need to cast to void *.
Isn't a direct assignment better than calling a memcpy on 4 bytes?

Reply:
If the 4 bytes of memory behind commandCode are not page-aligned, i.e. cross a page border, then the assignment may fail, if I remember correctly.

The thing about (void *), shall I change this ?

Reply:
If that was the case then all direct assignments of ints would have to be change to memcpy.
I think direct assignment is good here and we don't need the tmp variable (which BTW can also be unaligned).

Reply:
The thing is that you may not cast a uint8_t[] into a uint32_t, because the former is 1-byte-aligned, whereas the latter is 4-byte-aligned.

I may be wrong about this, but I thought that an uint32_t would never go over page-borders, whereas in a uint8_t[]-array different fields of the array can be on different pages... I have run something like this on some Coretex-M3 at some point... But hardly remember.

P.S. in x86 I think the CPU can handle cross-page-primitive types. But others cannot.

So yes, whenever a byte-array is cast into a multi-byte integer, these places need to be changed...

Also see https://wiki.sei.cmu.edu/confluence/display/c/EXP36-C.+Do+not+cast+pointers+into+more+strictly+aligned+pointer+types

Reply:
but now it is uint32_t across the board and has nothing to do with uint_8_t[4].
We don't do any conversion between uint_8_t[4] and uint32_t now right?

Reply:
`(void *)&tmp` basically goes from 4-byte-aligned to unaligned.

`*(TPM2_CC *)commandCode` used to go from unaligned to 4-byte-aligned, which was bad...

but now we are ok

Reply:
So do you want to update it to:
*(TPM2_CC *)commandCode = HOST_TO_BE_32(ctx->commandCode);
?

Reply:
No, `*(TPM2_CC *)commandCode` is something we must not do.

It breaks on certain platforms and it is against SEI recommendations.
I stand with my proposed patch.

Reply:
For my own education, could you please point me to the SEI recommendations that says that we should not do direct variable assignment.

Reply:
This one: https://wiki.sei.cmu.edu/confluence/display/c/EXP36-C.+Do+not+cast+pointers+into+more+strictly+aligned+pointer+types
Note especially EXP36-C-EX1 on that page.

The thing is not about the assignment, it is about casting a uint8_t* to a uint32_t*, i.e. "casting to a more strictly aligned type". Though the examples do not exactly match.

Also the second example here: https://wiki.sei.cmu.edu/confluence/display/c/EXP39-C.+Do+not+access+a+variable+through+a+pointer+of+an+incompatible+type
(Though this does not result in CPU-Exceptions, so I value the first one higher)